helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -1305,6 +1330,43 @@ String getMessageTarget(String resourceName, String partitionName) {
     return String.format("%s_%s", resourceName, partitionName);
   }
 
+  private void changeParticipantStatus(String instanceName,
+      LiveInstance.LiveInstanceStatus toStatus, HelixManager manager) {
+    HelixDataAccessor accessor = manager.getHelixDataAccessor();
+    String sessionId = manager.getSessionId();
+    String path = accessor.keyBuilder().liveInstance(instanceName).getPath();
+
+    if (LiveInstance.LiveInstanceStatus.PAUSED.equals(toStatus)) {
+      // Entering freeze mode, update live instance status
+      boolean success = accessor.getBaseDataAccessor().update(path, record -> {
+        record.setEnumField(LiveInstance.LiveInstanceProperty.STATUS.name(), toStatus);
+        return record;
+      }, AccessOption.EPHEMERAL);
+      if (success) {
+        _freezeSessionId = sessionId;
+        _liveInstanceStatus = toStatus;
+      }
+    } else if (LiveInstance.LiveInstanceStatus.NORMAL.equals(toStatus)) {
+      // Exiting freeze mode
+      // session changed, should call state model syncState()
+      if (_freezeSessionId != null && !_freezeSessionId.equals(sessionId)) {
+        syncFactoryState();
+        PropertyKey.Builder keyBuilder = new PropertyKey.Builder(manager.getClusterName());
+        ParticipantManager
```

Comment:
As mentioned above, this carry should not change the state it has for that state model.